pkg/controller/virtualmachineimage/virtualmachineimage_controller.go

```diff
@@ -126,5 +113,43 @@ func (r *ReconcileVirtualMachineImage) validateVirtualMachineImageSpec() error {
 	if !found {
 		return goerrors.New("storage request in pvc is missing")
 	}
+	src := r.getSource()
+	if src == "" {
+		return goerrors.New("no source set for vmim")
+	} else if src == sourceTypeVolume {
+		// check node name is not empty when source is hostPath volume
+		if r.vmi.Spec.Source.Volume.HostPath != nil && r.vmi.Spec.NodeName == "" {
+			return goerrors.New("when the source is hostPath volume, the node name must not be empty")
+		}
+	}
 	return nil
 }
+
+func (r *ReconcileVirtualMachineImage) reconcileVirtualMachineImage() error {
+	// If the pvc doesn't exist, create a pvc and update vmim's status to creating
+	if err := r.syncPvc(); err != nil {
+		return err
+	}
+	if r.getSource() != sourceTypeVolume {
+		if err := r.syncScratchPvc(); err != nil {
+			return err
+		}
+	}
+	if err := r.syncImporterPod(); err != nil {
+		return err
+	}
+	if err := r.syncSnapshot(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ReconcileVirtualMachineImage) getSource() string {
+	if r.vmi.Spec.Source.HTTP != "" {
+		return sourceTypeHTTP
+	} else if r.vmi.Spec.Source.Volume != nil {
+		return sourceTypeVolume
+	} else {
+		return ""
```

Comment:
this return should be handled with error

Reply:
no resource error handled in line 116
`if src := r.getSource(); src == "" {return goerrors.New("no source set for vmim")`}

Reply:
there is no meaning that 'empty' resource. is it valid state? no, it never be used.
not important the error is handled eventually, but what is the meaning of the function.